src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialects.java

```diff
@@ -0,0 +1,198 @@
+   * database connectivity mechanism that may be supported by one or more drivers. The contents and
+   * syntax of the {@code subname} will depend on the subprotocol.
+   *
+   * <p>The subprotocol will be in group 1, and the subname will be in group 2.
+   */
+  private static final Pattern PROTOCOL_PATTERN = Pattern.compile("jdbc:([^:]+):(.*)?");
+  private static final Logger LOG = LoggerFactory.getLogger(DatabaseDialects.class);
+  // Sort lexicographically to maintain order
+  private static final ConcurrentMap<String, DatabaseDialectProvider> REGISTRY = new
+      ConcurrentSkipListMap<>();
+
+  static {
+    loadAllDialects();
+  }
+
+  private static void loadAllDialects() {
+    LOG.debug("Searching for and loading all JDBC source dialects on the classpath");
+    final AtomicInteger count = new AtomicInteger();
+    AccessController.doPrivileged(new PrivilegedAction<Void>() {
+      public Void run() {
+        ServiceLoader<DatabaseDialectProvider> loadedDialects = ServiceLoader.load(
+            DatabaseDialectProvider.class);
+        Iterator<DatabaseDialectProvider> dialectIterator = loadedDialects.iterator();
+        try {
+          while (dialectIterator.hasNext()) {
+            try {
+              DatabaseDialectProvider provider = dialectIterator.next();
+              register(provider);
+              count.incrementAndGet();
+              LOG.debug("Found '{}' provider {}", provider, provider.getClass());
+            } catch (Throwable t) {
+              LOG.debug("Skipping dialect provider after error while loading", t);
+            }
+          }
+        } catch (Throwable t) {
+          LOG.debug("Error loading dialect providers", t);
+        }
+        return null;
+      }
+    });
+    LOG.debug("Registered {} source dialects", count.get());
+  }
+
+  /**
+   * Register a {@link DatabaseDialectProvider} of {@link DatabaseDialect}s.
+   *
+   * @param provider the dialect provider; may not be null
+   */
+  public static void register(DatabaseDialectProvider provider) {
+    REGISTRY.put(provider.getClass().getName(), provider);
+  }
+
+  /**
+   * Find the {@link DatabaseDialect} that has the highest {@link
+   * DatabaseDialectProvider#score(JdbcUrlInfo) score} for the supplied JDBC URL and Connection,
+   * and return a new instance of that dialect. Note that the DatabaseDialect needs to be
+   * {@link DatabaseDialect#close() closed}.
+   *
+   * @param jdbcUrl the JDBC connection URL; may not be null
+   * @param config  the connector configuration used to create the dialect; may not be null
+   * @return the {@link DatabaseDialect} instance with the greatest score; never null, but possibly
+   *     the {@link DatabaseDialect default DatabaseDialect}
+   * @throws ConnectException if there is a problem with the JDBC URL
+   */
+  public static DatabaseDialect findBestFor(
+      String jdbcUrl,
+      AbstractConfig config
+  ) throws ConnectException {
```

Comment:
This class and in particular this method is the starting point for using dialects. Given a JDBC URL and a source or sink connector configuration, this method iterates through all of the registered `DatabaseDialectProvider` instances and has each provider score the given JDBC URL. The provider with the highest score will be used as a factory to create a new `DatabaseDialect` instance with the supplied connector configuration.

Although `DatabaseDialectProvider` implementations can be registered manually, this class uses Java's Service Provider mechanism to automatically find and register all `DatabaseDialectProvider` implementations that are accessible via this class' classloader. That means that anyone can implement a custom `DatabaseDialectProvider`, build a properly-constructed JAR file w/ provider manifest file, and add that JAR file to the JDBC connector plugin directory. (See the JavaDoc for the `DatabaseDialectProvider` interface for specifics.)